pkg/handlers/rpm.go

```diff
@@ -22,63 +22,66 @@ func newRPMHandler() *rpmHandler {
 
 // HandleFile processes RPM formatted files. Further implementation is required to appropriately
 // handle RPM specific archive operations.
-func (h *rpmHandler) HandleFile(ctx logContext.Context, input fileReader) (chan []byte, error) {
-	archiveChan := make(chan []byte, defaultBufferSize)
+func (h *rpmHandler) HandleFile(ctx logContext.Context, input fileReader) chan DataOrErr {
+	dataOrErrChan := make(chan DataOrErr, defaultBufferSize)
 
 	if feature.ForceSkipArchives.Load() {
-		close(archiveChan)
-		return archiveChan, nil
+		close(dataOrErrChan)
+		return dataOrErrChan
 	}
 
 	go func() {
-		ctx, cancel := logContext.WithTimeout(ctx, maxTimeout)
-		defer cancel()
-		defer close(archiveChan)
+		defer close(dataOrErrChan)
 
 		// Update the metrics for the file processing.
 		start := time.Now()
 		var err error
 		defer func() {
-			h.measureLatencyAndHandleErrors(start, err)
+			h.measureLatencyAndHandleErrors(ctx, start, err, dataOrErrChan)
 			h.metrics.incFilesProcessed()
 		}()
 
 		// Defer a panic recovery to handle any panics that occur during the RPM processing.
 		defer func() {
 			if r := recover(); r != nil {
-				// Return the panic as an error.
+				var panicErr error
 				if e, ok := r.(error); ok {
-					err = e
+					panicErr = e
 				} else {
-					err = fmt.Errorf("panic occurred: %v", r)
+					panicErr = fmt.Errorf("panic occurred: %v", r)
+				}
+				dataOrErrChan <- DataOrErr{
+					Data: nil,
+					Err:  fmt.Errorf("%w: panic error: %w", ErrCriticalProcessing, panicErr),
 				}
-				ctx.Logger().Error(err, "Panic occurred when reading rpm archive")
 			}
 		}()
 
 		var rpm *rpmutils.Rpm
 		rpm, err = rpmutils.ReadRpm(input)
 		if err != nil {
-			ctx.Logger().Error(err, "error reading RPM")
```

Comment:
Are these not real errors?

Reply:
I understand why you thought I removed them. 😢 In the `defer func()`, we call `.measureLatencyAndHandleErrors`, which uses the shadowed `err`. This eventually gets logged when we consume from `dataOrErrChan`. I can avoid shadowing by explicitly sending `err` to the channel.